File: examples/PHP/mdbroker.php
<?php
/*
 * Majordomo Protocol broker
 * A minimal implementation of http://rfc.zeromq.org/spec:7 and spec:8
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */
include_once 'zmsg.php';
include_once 'mdp.php';

//  We'd normally pull these from config data
define("HEARTBEAT_LIVENESS", 3);    //  3-5 is reasonable
define("HEARTBEAT_INTERVAL", 2500); //  msecs
define("HEARTBEAT_EXPIRY", HEARTBEAT_INTERVAL * HEARTBEAT_LIVENESS);

/* Main broker work happens here */
$verbose = $_SERVER['argc'] > 1 && $_SERVER['argv'][1] == '-v';
$broker = new Mdbroker($verbose);
$broker->bind("tcp://*:5555");
$broker->listen();

class mdbroker
{
    private $ctx;               //  Our context
    private $socket;            //  Socket for clients & workers
    private $endpoint;          //  Broker binds to this endpoint

    private $services = array();          //  Hash of known services
    private $workers = array();           //  Hash of known workers
    private $waiting = array();           //  List of waiting workers

    private $verbose = false;   //  Print activity to stdout

    //  Heartbeat management
    private $heartbeat_at;  //  When to send HEARTBEAT

    /**
     * Constructor
     *
     * @param boolean $verbose
     */
    public function __construct($verbose = false)
    {
        $this->ctx = new ZMQContext();
        $this->socket = new ZMQSocket($this->ctx, ZMQ::SOCKET_ROUTER);
        $this->verbose = $verbose;
        $this->heartbeat_at = microtime(true) + (HEARTBEAT_INTERVAL/1000);
    }

    /**
     * Bind broker to endpoint, can call this multiple time
     * We use a single socket for both clients and workers.
     *
     * @param string $endpoint
     */
    public function bind($endpoint)
    {
        $this->socket->bind($endpoint);
        if ($this->verbose) {
            printf("I: MDP broker/0.1.1 is active at %s %s", $endpoint, PHP_EOL);
        }
    }

    /**
     * This is the main listen and process loop
     */
    public function listen()
    {
        $read = $write = array();

        //  Get and process messages forever or until interrupted
        while (true) {
            $poll = new ZMQPoll();
            $poll->add($this->socket, ZMQ::POLL_IN);

            $events = $poll->poll($read, $write, HEARTBEAT_INTERVAL);

            //  Process next input message, if any
            if ($events) {
                $zmsg = new Zmsg($this->socket);
                $zmsg->recv();
                if ($this->verbose) {
                    echo "I: received message:", PHP_EOL, $zmsg->__toString();
                }

                $sender = $zmsg->pop();
                $empty = $zmsg->pop();
                $header = $zmsg->pop();

                if ($header == MDPC_CLIENT) {
                    $this->client_process($sender, $zmsg);
                } elseif ($header == MDPW_WORKER) {
                    $this->worker_process($sender, $zmsg);
                } else {
                    echo "E: invalid message", PHP_EOL, $zmsg->__toString();
                }
            }

            //  Disconnect and delete any expired workers
            //  Send heartbeats to idle workers if needed
            if (microtime(true) > $this->heartbeat_at) {
                $this->purge_workers();
                foreach ($this->workers as $worker) {
                    $this->worker_send($worker, MDPW_HEARTBEAT, NULL, NULL);
                }
                $this->heartbeat_at = microtime(true) + (HEARTBEAT_INTERVAL/1000);
            }
        }
    }

    /**
     * Delete any idle workers that haven't pinged us in a while.
     * We know that workers are ordered from oldest to most recent.
     */
    public function purge_workers()
    {
        foreach ($this->waiting as $id => $worker) {
            if (microtime(true) < $worker->expiry) {
                break;      //  Worker is alive, we're done here
            }
            if ($this->verbose) {
                printf("I: deleting expired worker: %s %s",
                    $worker->identity, PHP_EOL);
            }
            $this->worker_delete($worker);
        }
    }

    /**
     * Locate or create new service entry
     *
     * @param  string   $name
     * @return stdClass
     */
    public function service_require($name)
    {
        $service = isset($this->services[$name]) ? $this->services[$name] : NULL;
        if ($service == NULL) {
            $service = new stdClass();
            $service->name = $name;
            $service->requests = array();
            $service->waiting = array();
            $this->services[$name] = $service;
        }

        return $service;
    }

    /**
     * Dispatch requests to waiting workers as possible
     *
     * @param type $service
     * @param type $msg
     */
    public function service_dispatch($service, $msg)
    {
        if ($msg) {
            $service->requests[] = $msg;
        }

        $this->purge_workers();

        while (count($service->waiting) && count($service->requests)) {
            $worker = array_shift($service->waiting);
            $msg = array_shift($service->requests);
            $this->worker_send($worker, MDPW_REQUEST, NULL, $msg);
        }
    }

    /**
     * Handle internal service according to 8/MMI specification
     *
     * @param string $frame
     * @param Zmsg   $msg
     */
    public function service_internal($frame, $msg)
    {
        if ($frame == "mmi.service") {
            $name = $msg->last();
            $service = $this->services[$name];
            $return_code = $service && $service->workers ? "200" : "404";
        } else {
            $return_code = "501";
        }

        $msg->set_last($return_code);

        //  Remove & save client return envelope and insert the
        //  protocol header and service name, then rewrap envelope
        $client = $msg->unwrap();
        $msg->push($frame);
        $msg->push(MDPC_CLIENT);
        $msg->wrap($client, "");
        $msg->set_socket($this->socket)->send();
    }

    /**
     * Creates worker if necessary
     *
     * @param  string   $address
     * @return stdClass
     */
    public function worker_require($address)
    {
        $worker = isset($this->workers[$address]) ? $this->workers[$address] : NULL;
        if ($worker == NULL) {
            $worker = new stdClass();
            $worker->identity = $address;
            $worker->address = $address;
            if ($this->verbose) {
                printf("I: registering new worker: %s %s", $address, PHP_EOL);
            }
            $this->workers[$address] = $worker;
        }

        return $worker;
    }

    /**
     * Remove a worker
     *
     * @param stdClass $worker
     * @param boolean  $disconnect
     */
    public function worker_delete($worker, $disconnect = false)
    {
        if ($disconnect) {
            $this->worker_send($worker, MDPW_DISCONNECT, NULL, NULL);
        }

        if (isset($worker->service)) {
            worker_remove_from_array($worker, $worker->service->waiting);
            $worker->service->workers--;
        }
        worker_remove_from_array($worker, $this->waiting);
        unset($this->workers[$worker->identity]);
    }

    private function worker_remove_from_array($worker, &$array)
    {
        $index = array_search($worker, $array);
        if ($index !== false) {
            unset($array[$index]);
        }
    }

    /**
     * Process message sent to us by a worker
     *
     * @param string $sender
     * @param Zmsg   $msg
     */
    public function worker_process($sender, $msg)
    {
        $command = $msg->pop();
        $worker_ready = isset($this->workers[$sender]);
        $worker = $this->worker_require($sender);
        if ($command == MDPW_READY) {
            if ($worker_ready) {
                $this->worker_delete($worker, true); //  Not first command in session
            } else if(strlen($sender) >= 4      // Reserved service name
                    && substr($sender, 0, 4) == 'mmi.') {
                $this->worker_delete($worker, true);
            } else {
                //  Attach worker to service and mark as idle
                $service_frame = $msg->pop();
                $worker->service = $this->service_require($service_frame);
                $worker->service->workers++;
                $this->worker_waiting($worker);
            }
        } elseif ($command == MDPW_REPLY) {
            if ($worker_ready) {
                //  Remove & save client return envelope and insert the
                //  protocol header and service name, then rewrap envelope.
                $client = $msg->unwrap();
                $msg->push($worker->service->name);
                $msg->push(MDPC_CLIENT);
                $msg->wrap($client, "");
                $msg->set_socket($this->socket)->send();
                $this->worker_waiting($worker);
            } else {
                $this->worker_delete($worker, true);
            }
        } elseif ($command == MDPW_HEARTBEAT) {
            if ($worker_ready) {
                $worker->expiry = microtime(true) + (HEARTBEAT_EXPIRY/1000);
            } else {
                $this->worker_delete($worker, true);
            }
        } elseif ($command == MDPW_DISCONNECT) {
            $this->worker_delete($worker, true);
        } else {
            echo "E: invalid input message", PHP_EOL, $msg->__toString();
        }
    }

    /**
     * Send message to worker
     *
     * @param stdClass $worker
     * @param string   $command
     * @param mixed    $option
     * @param Zmsg     $msg
     */
    public function worker_send($worker, $command, $option, $msg)
    {
        $msg = $msg ? $msg : new Zmsg();
        //  Stack protocol envelope to start of message
        if ($option) {
            $msg->push($option);
        }
        $msg->push($command);
        $msg->push(MDPW_WORKER);

        //  Stack routing envelope to start of message
        $msg->wrap($worker->address, "");

        if ($this->verbose) {
            printf("I: sending %s to worker %s",
               $command, PHP_EOL);
            echo $msg->__toString();
        }

        $msg->set_socket($this->socket)->send();
    }

    /**
     * This worker is now waiting for work
     *
     * @param stdClass $worker
     */
    public function worker_waiting($worker)
    {
        //  Queue to broker and service waiting lists
        $this->waiting[] = $worker;
        $worker->service->waiting[] = $worker;
        $worker->expiry = microtime(true) + (HEARTBEAT_EXPIRY/1000);
        $this->service_dispatch($worker->service, NULL);
    }

    /**
     * Process a request coming from a client
     *
     * @param string $sender
     * @param Zmsg   $msg
     */
    public function client_process($sender, $msg)
    {
        $service_frame = $msg->pop();
        $service = $this->service_require($service_frame);

        //  Set reply return address to client sender
        $msg->wrap($sender, "");
        if (substr($service_frame, 0, 4) == 'mmi.') {
            $this->service_internal($service_frame, $msg);
        } else {
            $this->service_dispatch($service, $msg);
        }
    }
}

File: examples/Go/mdbroker.go
//  Majordomo Protocol broker
//  A minimal C implementation of the Majordomo Protocol as defined in
//  http://rfc.zeromq.org/spec:7 and http://rfc.zeromq.org/spec:8.
// 
//  To run this example, you may need to run multiple *.go files as below
//  go run mdp.go zhelpers.go mdbroker.go [-v]
//
//  Author: iano <scaly.iano@gmail.com>
//  Based on C & Python example

package main

import (
	"encoding/hex"
	zmq "github.com/alecthomas/gozmq"
	"log"
	"os"
	"time"
)

const (
	INTERNAL_SERVICE_PREFIX = "mmi."
	HEARTBEAT_INTERVAL      = 2500 * time.Millisecond
	HEARTBEAT_EXPIRY        = HEARTBEAT_INTERVAL * HEARTBEAT_LIVENESS
)

type Broker interface {
	Close()
	Run()
}

type mdbWorker struct {
	identity string     //  Hex Identity of worker
	address  []byte     //  Address to route to
	expiry   time.Time  //  Expires at this point, unless heartbeat
	service  *mdService //  Owning service, if known
}

type mdService struct {
	broker   Broker
	name     string
	requests [][][]byte //  List of client requests
	waiting  *ZList     //  List of waiting workers
}

type mdBroker struct {
	context     *zmq.Context          //  Context
	heartbeatAt time.Time             //  When to send HEARTBEAT
	services    map[string]*mdService //  Known services
	socket      *zmq.Socket           //  Socket for clients & workers
	waiting     *ZList                //  Idle workers
	workers     map[string]*mdbWorker //  Known workers
	verbose     bool                  //  Print activity to stdout
}

func NewBroker(endpoint string, verbose bool) Broker {
	context, _ := zmq.NewContext()
	socket, _ := context.NewSocket(zmq.ROUTER)
	socket.SetLinger(0)
	socket.Bind(endpoint)
	log.Printf("I: MDP broker/0.1.1 is active at %s\n", endpoint)
	return &mdBroker{
		context:     context,
		heartbeatAt: time.Now().Add(HEARTBEAT_INTERVAL),
		services:    make(map[string]*mdService),
		socket:      socket,
		waiting:     NewList(),
		workers:     make(map[string]*mdbWorker),
		verbose:     verbose,
	}
}

//  Deletes worker from all data structures, and deletes worker.
func (self *mdBroker) deleteWorker(worker *mdbWorker, disconnect bool) {
	if worker == nil {
		panic("Nil worker")
	}

	if disconnect {
		self.sendToWorker(worker, MDPW_DISCONNECT, nil, nil)
	}

	if worker.service != nil {
		worker.service.waiting.Delete(worker)
	}
	self.waiting.Delete(worker)
	delete(self.workers, worker.identity)
}

//  Dispatch requests to waiting workers as possible
func (self *mdBroker) dispatch(service *mdService, msg [][]byte) {
	if service == nil {
		panic("Nil service")
	}
	//  Queue message if any
	if len(msg) != 0 {
		service.requests = append(service.requests, msg)
	}
	self.purgeWorkers()
	for service.waiting.Len() > 0 && len(service.requests) > 0 {
		msg, service.requests = service.requests[0], service.requests[1:]
		elem := service.waiting.Pop()
		self.waiting.Remove(elem)
		worker, _ := elem.Value.(*mdbWorker)
		self.sendToWorker(worker, MDPW_REQUEST, nil, msg)
	}
}

// Process a request coming from a client.
func (self *mdBroker) processClient(sender []byte, msg [][]byte) {
	//  Service name + body
	if len(msg) < 2 {
		panic("Invalid msg")
	}
	service := msg[0]
	//  Set reply return address to client sender
	msg = append([][]byte{sender, nil}, msg[1:]...)
	if string(service[:4]) == INTERNAL_SERVICE_PREFIX {
		self.serviceInternal(service, msg)
	} else {
		self.dispatch(self.requireService(string(service)), msg)
	}
}

//  Process message sent to us by a worker.
func (self *mdBroker) processWorker(sender []byte, msg [][]byte) {
	//  At least, command
	if len(msg) < 1 {
		panic("Invalid msg")
	}

	command, msg := msg[0], msg[1:]
	identity := hex.EncodeToString(sender)
	worker, workerReady := self.workers[identity]
	if !workerReady {
		worker = &mdbWorker{
			identity: identity,
			address:  sender,
			expiry:   time.Now().Add(HEARTBEAT_EXPIRY),
		}
		self.workers[identity] = worker
		if self.verbose {
			log.Printf("I: registering new worker: %s\n", identity)
		}
	}

	switch string(command) {
	case MDPW_READY:
		//  At least, a service name
		if len(msg) < 1 {
			panic("Invalid msg")
		}
		service := msg[0]
		//  Not first command in session or Reserved service name
		if workerReady || string(service[:4]) == INTERNAL_SERVICE_PREFIX {
			self.deleteWorker(worker, true)
		} else {
			//  Attach worker to service and mark as idle
			worker.service = self.requireService(string(service))
			self.workerWaiting(worker)
		}
	case MDPW_REPLY:
		if workerReady {
			//  Remove & save client return envelope and insert the
			//  protocol header and service name, then rewrap envelope.
			client := msg[0]
			msg = append([][]byte{client, nil, []byte(MDPC_CLIENT), []byte(worker.service.name)}, msg[2:]...)
			self.socket.SendMultipart(msg, 0)
			self.workerWaiting(worker)
		} else {
			self.deleteWorker(worker, true)
		}
	case MDPW_HEARTBEAT:
		if workerReady {
			worker.expiry = time.Now().Add(HEARTBEAT_EXPIRY)
		} else {
			self.deleteWorker(worker, true)
		}
	case MDPW_DISCONNECT:
		self.deleteWorker(worker, false)
	default:
		log.Println("E: invalid message:")
		Dump(msg)
	}
}

//  Look for & kill expired workers.
//  Workers are oldest to most recent, so we stop at the first alive worker.
func (self *mdBroker) purgeWorkers() {
	now := time.Now()
	for elem := self.waiting.Front(); elem != nil; elem = self.waiting.Front() {
		worker, _ := elem.Value.(*mdbWorker)
		if worker.expiry.After(now) {
			break
		}
		self.deleteWorker(worker, false)
	}
}

//  Locates the service (creates if necessary).
func (self *mdBroker) requireService(name string) *mdService {
	if len(name) == 0 {
		panic("Invalid service name")
	}
	service, ok := self.services[name]
	if !ok {
		service = &mdService{
			name:    name,
			waiting: NewList(),
		}
		self.services[name] = service
	}
	return service
}

//  Send message to worker.
//  If message is provided, sends that message.
func (self *mdBroker) sendToWorker(worker *mdbWorker, command string, option []byte, msg [][]byte) {
	//  Stack routing and protocol envelopes to start of message and routing envelope
	if len(option) > 0 {
		msg = append([][]byte{option}, msg...)
	}
	msg = append([][]byte{worker.address, nil, []byte(MDPW_WORKER), []byte(command)}, msg...)

	if self.verbose {
		log.Printf("I: sending %X to worker\n", command)
		Dump(msg)
	}
	self.socket.SendMultipart(msg, 0)
}

//  Handle internal service according to 8/MMI specification
func (self *mdBroker) serviceInternal(service []byte, msg [][]byte) {
	returncode := "501"
	if string(service) == "mmi.service" {
		name := string(msg[len(msg)-1])
		if _, ok := self.services[name]; ok {
			returncode = "200"
		} else {
			returncode = "404"
		}
	}
	msg[len(msg)-1] = []byte(returncode)
	//  insert the protocol header and service name after the routing envelope
	msg = append(append(msg[2:], []byte(MDPC_CLIENT), service), msg[3:]...)
	self.socket.SendMultipart(msg, 0)
}

//  This worker is now waiting for work.
func (self *mdBroker) workerWaiting(worker *mdbWorker) {
	//  Queue to broker and service waiting lists
	self.waiting.PushBack(worker)
	worker.service.waiting.PushBack(worker)
	worker.expiry = time.Now().Add(HEARTBEAT_EXPIRY)
	self.dispatch(worker.service, nil)
}

func (self *mdBroker) Close() {
	if self.socket != nil {
		self.socket.Close()
	}
	self.context.Close()
}

//  Main broker working loop
func (self *mdBroker) Run() {
	for {
		items := zmq.PollItems{
			zmq.PollItem{Socket: self.socket, Events: zmq.POLLIN},
		}

		_, err := zmq.Poll(items, HEARTBEAT_INTERVAL)
		if err != nil {
			panic(err) //  Interrupted
		}

		if item := items[0]; item.REvents&zmq.POLLIN != 0 {
			msg, _ := self.socket.RecvMultipart(0)
			if self.verbose {
				log.Printf("I: received message:")
				Dump(msg)
			}

			sender := msg[0]
			header := msg[2]
			msg = msg[3:]

			if string(header) == MDPC_CLIENT {
				self.processClient(sender, msg)
			} else if string(header) == MDPW_WORKER {
				self.processWorker(sender, msg)
			} else {
				log.Println("E: invalid message:")
				Dump(msg)
			}
		}

		if self.heartbeatAt.Before(time.Now()) {
			self.purgeWorkers()
			for elem := self.waiting.Front(); elem != nil; elem = elem.Next() {
				worker, _ := elem.Value.(*mdbWorker)
				self.sendToWorker(worker, MDPW_HEARTBEAT, nil, nil)
			}
			self.heartbeatAt = time.Now().Add(HEARTBEAT_INTERVAL)
		}
	}
}

func main() {
	verbose := len(os.Args) >= 2 && os.Args[1] == "-v"
	broker := NewBroker("tcp://*:5555", verbose)
	defer broker.Close()

	broker.Run()
}
